### ISSUE TITLE
Sql Server RegisterOrUpdateReminder deadlock

### ISSUE THREAD

**Issue by mtdaniels** [dotnet/orleans#2429]
I'm seeing sql deadlocks occasionally when RegisterOrUpdateReminder is called.

We use reminders somewhat heavily to make grain work durable (if, for some reason, the grain is killed while performing work, the reminder will reactivate the grain, and work will continue).

I believe the upsert is creating the deadlock because the insert upgrades the lock scope from a single row to the key range.  So, two transactions are able to get row locks that are in the same range, but when they try to insert (because neither currently exist), they deadlock.

I saw this: https://github.com/dotnet/orleans/pull/2395 - but it doesn't look like that modified the reminders implementation.

I have worked around the deadlock issue by adding retries, but this is obviously just masking the underlying problem.

**Comment by sergeybykov** [comment]
@mtdaniels Did you mean to point to https://github.com/dotnet/orleans/pull/2395 rather?

I wonder what @veikkoeeva and @shayhatsor think.

**Comment by mtdaniels** [comment]
Indeed I did - sorry about that

**Comment by veikkoeeva** [comment]
@mtdaniels Do you mean reminders or storage deadlocks? It's entirely plausible reminders deadlock and it could be greatly mitigated (and likely removed altogether), there's no optimizations done to the queries and the table from what they originally were. The locking could be eased and it would greatly benefit using indexes to filter the where clauses.

**Comment by mtdaniels** [comment]
@veikkoeeva - reminders deadlock.

**Comment by veikkoeeva** [comment]
@mtdaniels Would it be OK to consider if you alter the table with some judicious use of indexes and possibly augment the queries and if it helps, make a PR here? :) I can work on this too, but I'm in a bind for about two weeks before I can take on this.

Indeed, the reminder queries are very plain, unoptimized and painful to watch. I apologize.

**Comment by veikkoeeva** [comment]
You can alter the tables practically on the fly. Though currently you have to restart silos if you alter the queries, since currently the silos reload queries only then.

**Comment by veikkoeeva** [comment]
@mtdaniels May I ask, how many remidners you have, by the way? If you hundreds of millions, we should do more extensive modifications. I think doable still, but adding a few simple indexes wouldn't be the most efficient way to go. :)

**Comment by mtdaniels** [comment]
Typically not very many exist at a single time (on the order of hundreds).  There is somewhat high throughput (1,500/minute).  The majority of them are deleted before they ever fire.

If two executions of RegisterOrUpdateReminder occur at the same time, land in the same key range, and result in insertions - then I believe you have a deadlock (because of the upsert).

**Comment by shayhatsor** [comment]
@mtdaniels, thanks for reporting this issue. Even though the reminders queries aren't as optimized as they might be, I was under the impression that the deadlock issues are behind us. Can you please provide the SQL Server error log? this will help us make sure this doesn't reoccur.

**Comment by veikkoeeva** [comment]
@shayhatsor All serialiazing locks are problematic. The problems is at https://github.com/dotnet/orleans/blob/master/src/OrleansSQLUtils/CreateOrleansTables_SqlServer.sql#L390. A mostly working fix would be quick too (examples at http://stackoverflow.com/questions/108403/solutions-for-insert-or-update-on-sql-server, say), but I wonder if we should make a quicker fix now and work a more scalable solution later.

The other problem one bumps into reminders when at sufficient scale is the queries aren't indexed and hence slow and use much more resources.

**Comment by shayhatsor** [comment]
@veikkoeeva, I believe that @mtdaniels 's specific use of reminders:
> We use reminders somewhat heavily to make grain work durable (if, for some reason, the grain is killed while performing work, the reminder will reactivate the grain, and work will continue).

 isn't the usual use case and that's why this just came up now (AFAIK). I suggest the following:
@mtdaniels - I think we might be able to suggest other Orleans mechanisms to achieve the same outcome. Please provide more info as to the cases you used reminders to activate 'killed' grains.
@veikkoeeva - Let's fix this once and for all. Starting with a test that reproduces the bug. There's no rush.

**Comment by mtdaniels** [comment]
@shayhatsor - you're probably right.  My guess is that most Orleans implementations do not regularly create / destroy reminders.

We have a worker grain that uses pub/sub to perform long-running work.  If the service is restarted / crashes / whatever, the reminder will reactivate the grain.  The grain uses its state to determine that it was running, it has not finished, and it should restart the work.  This guarantees the work will be completed (at least eventually), and the subscriber won't sit waiting indefinitely.

I know it is against best practices to wrap long running work in an Orleans grain; however, this approach works (for the most part).

**Comment by veikkoeeva** [comment]
@shayhatsor Yeah, we should fix this for good. Proper indexes, soft deletes, as `DELETE` is very disruptive on indexes. Then I maybe separate the Id from the type like in Storage, maybe arranging the table a bit like that in any event. Also maybe hashing the ID and type. I remember there was a case on ATS having enough reminders querying them from storage came prohibitevely slow, so maybe we should aim to solve this so it wouldn't be a problem (i.e. fast index look-ups and little maintenance).

@mtdaniels How quickly would you like to have a fix? You can unblock yourself by just modifying the script and/or table, but indeed and keep the input paramenter names and types and output columuns names and types.

**Comment by mtdaniels** [comment]
The deadlock issue is infrequent enough that adding a retry isn't terrible (for now).

I am finding, though, that my reminders regularly fail to unregister.  At first glance, it appears both GetReminder and UnregisterReminder timeout frequently.   I'll have to investigate that more after the holiday.

**Comment by amccool** [comment]
We just got a dead lock on the reminder table as well

using
select cmd,* from sys.sysprocesses
where blocked > 0

yields:

(gonna try to format this shortly)

cmd	spid	kpid	blocked	waittype	waittime	lastwaittype	waitresource	dbid	uid	cpu	physical_io	memusage	login_time	last_batch	ecid	open_tran	status	sid	hostname	program_name	hostprocess	cmd	nt_domain	nt_username	net_address	net_library	loginame	context_info	sql_handle	stmt_start	stmt_end	request_id
UPDATE          	87	8064	68	0x000E	3019	LCK_M_RS_U                      	KEY: 6:72057594041532416 (114367bccbd5)                                                                                                                                                                                                                         	6	5	0	9	3	2017-04-25 13:41:22.617	2017-04-25 13:41:22.617	0	2	suspended                     	0x9077550267FDF446A44EA9CC562048AF00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000	NCI-R5EAS01                                                                                                                     	.Net SqlClient Data Provider                                                                                                    	2464      	UPDATE          	                                                                                                                                	                                                                                                                                	42C824D26A77	TCP/IP      	orleans                                                                                                                         	0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000	0x0100060043B12F2A202A0CBADB00000000000000	412	1124	0

**Comment by ReubenBond** [comment]
Closing since this hasn't been reported in quite some time and might be fixed without noticing. If this still occurs, please open a new issue and reference this one.
